Src/ILGPU.Algorithms.Tests/MatrixTests.cs

```diff
@@ -0,0 +1,258 @@
+            {
+                for (var y = 0; y < rightColumns; y++)
+                {
+                    for (var z = 0; z < leftColumns; z++)
+                        result[x, y] += left[x, z] * right[z, y];
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Compute the transpose of a matrix in 2D.
+        /// </summary>
+        /// <param name="matrix">A MxN matrix</param>
+        /// <returns>The transpose of A, a NxM matrix</returns>
+        private static float[,] TransposeMatrix2D(float[,] matrix)
+        {
+            int rows = matrix.GetLength(0);
+            int columns = matrix.GetLength(1);
+            var transposed = new float[columns, rows];
+            for (var i = 0; i < rows; i++)
+            {
+                for (var j = 0; j < columns; j++)
+                    transposed[j, i] = matrix[i, j];
+            }
+
+            return transposed;
+        }
+
+        /// <summary>
+        /// Compares two matrices for equality.
+        /// </summary>
+        /// <param name="left">A MxN matrix (the actual matrix we got)</param>
+        /// <param name="right">A MxN matrix (the matrix we expected) </param>
+        /// <returns>True if the matrices are equal</returns>
+        private static void AssertMatrixEqual2D(
+            float[,] left,
+            float[,] right,
+            float eps = 1e-4f)
+        {
+            var leftRows = left.GetLength(0);
+            var leftColumns = left.GetLength(1);
+            var rightRows = right.GetLength(0);
+            var rightColumns = right.GetLength(1);
+
+            if (leftRows != rightRows || leftColumns != rightColumns)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(right),
+                    $"Matrix dimensions {leftRows}x{leftColumns} and " +
+                    $"{rightRows}x{rightColumns} do not match");
+            }
+
+            for (var i = 0; i < leftRows; i++)
+            {
+                for (var j = 0; j < leftColumns; j++)
+                {
+                    Assert.True(
+                        Math.Abs(left[i, j] - right[i, j]) < eps,
+                        "Matrix result not equal");
+                }
+            }
+        }
+
+        #endregion
+
+        [SkippableTheory()]
+        [MemberData(nameof(DimensionsData))]
+        public void MatrixMultiplySparse(
+            int length,
+            int baseLength,
+            int seed,
+            float sparsityProbability)
+        {
+            Skip.If(length > 512);
+            
+            using var stream = Accelerator.CreateStream();
+            
+            // Our starting dense matrix
+            var denseMatrix = new float[baseLength, length];
+
+            // Setup sparse 2D matrix
+            var sparseMatrix = new float[length, length];
+            var sparseMaskMatrix = new float[length, length];
```

Comment:
I would recommend making the shape of sparseMatrix be [length2, length] to more explicitly test the transpose and dimension matching.